### ISSUE TITLE
[OrionMongoSink] The sink puts "/" in collections' names

### ISSUE THREAD

**Issue by redss** [telefonicaid/fiware-cygnus#586]
Hello,

We're using Cygnus to store historical data from Orion to MongoDB. To determine collection name, ``OrionMongoSink`` uses, among others, Orion's service path. Service paths in Orion always have at least one ``/`` character in them. This works fine until you want to backup the database.

On Linux with MongoDB 2.6.11, ``mongodump`` gives following errors:

```
Cannot dump parka_def_serv.parka/bays. Collection has '/' or null in the collection name.
Cannot dump parka_def_serv.parka/relaynodes. Collection has '/' or null in the collection name.
Cannot dump parka_def_serv.parka/datacollectors. Collection has '/' or null in the collection name.
```

On Windows with MongoDB 3.07 it's even worse – ``mongodump`` interprets ``/`` in collections' names as directories, which in consequence creates backups that cannot be restored by ``mongorestore``, as it doesn't know what to do with these directories.

The issue is mentioned on MongoDB's Jira: [SERVER-4413 Mongodump segfaults on certain collection names](https://jira.mongodb.org/browse/SERVER-4413). I think the conclusion is that collections shouldn't have ``/`` in their names, as it can cause problems in the future.

I guess, that the problem lies in [this line](https://github.com/telefonicaid/fiware-cygnus/blob/master/src/main/java/com/telefonica/iot/cygnus/sinks/OrionMongoSink.java#L66).

One solution would be is to use [MongoDB namespaces](https://docs.mongodb.org/manual/faq/developers/#what-is-a-namespace-in-mongodb) instead, or at least make character joining service paths configurable.

Best Regards,
Ryszard Tarajkowski

**Comment by frbattid** [comment]
Hi Ryzzard,

the '/' character was added to the collection name since we did not consider it as a restriction according to:
https://docs.mongodb.org/v2.6/reference/limits/#Restriction-on-Collection-Names 

Nevertheless, I've tested your scenario (dumping mongo data) and you are right, the '/ character is not accepted...

**Comment by gtorodelvalle** [comment]
In the line of what @frbattid mentions and although I guess you have already considered it, what about renaming the collections using https://docs.mongodb.org/v2.6/reference/command/renameCollection/ and substituting the `/` by `%2F` for example using a URL encoding as explained at http://www.freeformatter.com/url-encoder.html#url-encoding-explained before "mongodumping" them? Of course, the collection names should be decoded back when restoring... :)

A script could easily automate this process... ;) 

Just my 2 cents... :)

**Comment by frbattid** [comment]
I think it is a good solution to provide the kind of scripts @gtorodelvalle proposes, being the process: encoding, dumping and finally decoding. This will avoid changing the way the collection names are created and breaking backward compatibility.

**Comment by frbattid** [comment]
Of course, this applies to `OrionSTHSink` as well.

**Comment by frbattid** [comment]
This issue should be fixed with the new encoding mechanism available from release/1.3.0, that encodes the slash as `x2f00`. Thus, closin.
